apps/prairielearn/assets/scripts/instructorAssessmentInstancesClient.tsx

```diff
@@ -401,93 +415,235 @@ onDocumentReady(() => {
+    let explanation = '';
+    if (action === 'set_total') {
+      explanation =
+        'Updating the total time limit will set the given amount of time for the assessment based on when the assessment was started.';
+    } else if (action === 'set_rem') {
+      explanation =
+        'Updating the time remaining will set the given amount of time for the assessment based on the current time.';
+    } else if (action === 'set_exact') {
+      explanation = 'This will set the exact closing time for the assessment.';
+    } else if (action === 'add') {
+      explanation = 'This will add the given amount of time to the remaining time limit.';
+    } else if (action === 'subtract') {
+      explanation = 'This will subtract the given amount of time from the remaining time limit.';
+    } else if (action === 'remove') {
+      explanation = 'This will remove the time limit and the assessment will remain open.';
+    } else if (action === 'expire') {
+      explanation = 'This will expire the time limit and the assessment will be closed.';
+    }
+    return <small class="form-text text-muted">{explanation}</small>;
+  }
+
+  function TimeLimitEditPopover({
+    row,
+  }: {
+    row: {
+      action: string;
+      assessment_instance_id: number;
+      date: string;
+      has_closed_instance: boolean;
+      has_open_instance: boolean;
+      total_time: string;
+      total_time_sec: number;
+      time_remaining: string;
+      time_remaining_sec: number;
+      open: boolean;
+    };
+  }) {
+    const [form, setForm] = useState({
+      action: 'add' as TimeLimitAction,
+      time_add: 5,
+      date: Temporal.Now.zonedDateTimeISO(timezone).toPlainDateTime().toString().slice(0, 16),
+      reopen_closed: false,
+      reopen_without_limit: true,
+    });
+    const showTimeLimitOptions =
+      row.action === 'set_time_limit_all' || row.open || !form.reopen_without_limit;
+
+    function updateFormState<T extends keyof typeof form>(key: T, value: (typeof form)[T]) {
+      setForm({
+        ...form,
+        [key]: value,
+      });
+    }
+
+    function proposedClosingTime() {
+      const startDate = row.date;
+      const totalTime = Math.round(row.total_time_sec);
+
+      let temporalDate = Temporal.Instant.from(startDate).toZonedDateTimeISO(timezone);
+      if (form.action === 'set_total') {
+        temporalDate = temporalDate.add({ minutes: form.time_add });
+      } else if (form.action === 'set_rem') {
+        temporalDate = Temporal.Now.zonedDateTimeISO(timezone).add({ minutes: form.time_add });
+      } else if (form.action === 'add') {
+        temporalDate = temporalDate.add({ seconds: totalTime }).add({ minutes: form.time_add });
+      } else if (form.action === 'subtract') {
+        temporalDate = temporalDate
+          .add({ seconds: totalTime })
+          .subtract({ minutes: form.time_add });
+      }
+
+      return formatDate(new Date(temporalDate.toInstant().toString()), timezone);
```

Comment:
This is a more conventional way to convert from a `Temporal.Instant` to a `Date`, see https://tc39.es/proposal-temporal/docs/instant.html#interoperability-with-date:

```suggestion
      return formatDate(new Date(temporalDate.epochMilliseconds), timezone);
```

Reply:
Got it. Updated this locally in [04498d3](https://github.com/PrairieLearn/PrairieLearn/pull/10800/commits/04498d322330eb9d5e4dfc31e76260a6d9709196).